File: advent-of-code-2024/day5.py
with open('input') as f:
   data = f.read().split('\n\n')
   rules = {}
   for rule in data[0].split('\n'):
      n1, n2 = map(int, rule.split('|'))
      if n1 in set(rules.keys()): rules[n1].append(n2)
      else: rules[n1] = [n2]

   orderings = []
   for ordering in data[1].split('\n'):
      orderings.append([int(page) for page in ordering.split(',')])

part1 = 0

rightOrders = []
wrongOrders = []
for eachOrdering in orderings:
   rightOrder = True
   for i in range(len(eachOrdering)):
      for eachPage in eachOrdering[i+1:]:
         if eachOrdering[i] not in set(rules.keys()) or eachPage not in rules[eachOrdering[i]]: 
            rightOrder = False
            break
      if not rightOrder: break

   if rightOrder:  rightOrders.append(eachOrdering) 
   else: wrongOrders.append(eachOrdering)    

print("part 1 : ", sum([rightOrder[len(rightOrder) // 2] for rightOrder in rightOrders]))

correctedOrders = []
for wrongOrder in wrongOrders:
   correctedOrder = list(wrongOrder)
   for i in range(len(wrongOrder)):
      count = 0
      for page in wrongOrder[:i] + wrongOrder[i+1:]:
         if wrongOrder[i] in set(rules.keys()) and page in rules[wrongOrder[i]]: count += 1

      correctedOrder[count * -1 - 1] = wrongOrder[i]

   correctedOrders.append(correctedOrder)  

print("part2 : ", sum([correctOrder[len(correctOrder) // 2] for correctOrder in correctedOrders]))
      

   
    

File: advent-of-code-2024/cpp/day5.cpp
# include <iostream>
# include <fstream>
# include <vector>
# include <unordered_map>
# include <sstream>

# define NUM(s) ((s[0] - '0') * 10 + (s[1] - '0'))

using namespace std;

vector<string> split(string str, char at){
    vector<string> tokens;
    string token;
    istringstream stream(str);

    while(getline(stream, token, at)) tokens.push_back(token);

    return tokens;
}

vector<int> getNewOrdering(vector<int> & ordering, unordered_map<int, vector<int> > & rules){
    vector<int> newOrdering = ordering;
    for(int page1=0;page1 < ordering.size(); page1++){
        int count = 0;
        for(int page2=0; page2 < ordering.size(); page2++){
            if(page1 == page2) continue;

            if(find(rules[ordering[page1]].begin(), rules[ordering[page1]].end(), ordering[page2]) != rules[ordering[page1]].end()) {
                count++;
            }
        }
        newOrdering[ordering.size()-count-1] = ordering[page1];
    }
    return newOrdering;
}

int main(){
    ifstream file("../input");

    string str;
    unordered_map<int, vector<int> > rules;
    vector<vector<int> > orderings;

    while(file >> str){
        if(str.length() == 5){
            rules[NUM(str)].push_back(NUM(str.substr(3, 2)));
        } else {
            vector<int> ordering;

            for(auto s: split(str, ',')){
                ordering.push_back(NUM(s));
            }
            orderings.push_back(ordering);
        }
    }
    
    int part1 = 0;
    int part2 = 0;

    for(auto ordering: orderings){
        bool rightOrdering = true;
        for(int page1=0;page1 < ordering.size(); page1++){
            
            for(int page2=page1+1; page2 < ordering.size(); page2++){
                if(find(rules[ordering[page1]].begin(), rules[ordering[page1]].end(), ordering[page2]) == rules[ordering[page1]].end()) {
                    rightOrdering=false;
                    vector<int> newOrdering = getNewOrdering(ordering, rules);
                    part2 += newOrdering[newOrdering.size() / 2];
                    break;
                }
            }
            if(!rightOrdering) break;
        }

        if(rightOrdering) part1 += ordering[ordering.size() / 2];
    }
    
    cout << "part1 : " << part1 << endl;
    cout << "part2 : " << part2 << endl;

}